File: tools/buildFirefox.py
import os
import sys
import glob
import distutils.dir_util
import distutils.file_util
import json

sdkDir = "addon-sdk-1.17"
firefoxDir = "ff_tgarmory"
chromeDir = "tgarmory"

def run():
	print("Update data")
	print(os.getcwd())

	copyDirs = ["lib", "css", "images"]
	for cDir in copyDirs :
		distutils.dir_util.copy_tree(cDir, os.path.join("..", firefoxDir, "data", cDir))
	distutils.file_util.copy_file("tgarmory.js", os.path.join("..", firefoxDir, "data"))

	print("Update version")
	with open(os.path.join("manifest.json"), 'r') as content_file:
		manifest = json.load(content_file)
	with open(os.path.join("..", firefoxDir, "package.json"), 'r') as content_file:
		package = json.load(content_file)
	package["version"] = manifest["version"]
	with open(os.path.join("..", firefoxDir, "package.json"), 'w') as content_file:
		json.dump(package, content_file, indent=2)

	print("Update main.js")
	mainPrefix = '\n'.join([
		'// Import the page-mod API',
		'var pageMod = require("sdk/page-mod");',
		'// Import the self API',
		'var self = require("sdk/self");',
		'pageMod.PageMod({',
		'  include: ["*.thegeekcrusade-serveur.com"],',
		'  contentScriptFile: [',])
	mainMid = '\n'.join([
		'',
		'  ],',
		'  contentScriptOptions: {'])
	mainSuffix = '\n'.join([
		'',
		'  },',
		'});'])

	main = mainPrefix
	for js in manifest["content_scripts"][0]["js"] :
		main += "\n    self.data.url(\"%s\")," % js
	main = main.rstrip(",") + mainMid
	if "css" in manifest["content_scripts"][0]:
		for css in manifest["content_scripts"][0]["css"] :
			main += "\n    \"%s\": self.data.url(\"%s\")," % (css, css)
		main = main.rstrip(",")
	if "web_accessible_resources" in manifest:
		for res in manifest["web_accessible_resources"] :
			main += "\n    \"%s\": self.data.url(\"%s\")," % (res, res)
		main = main.rstrip(",")
	main = main + mainSuffix

	with open(os.path.join("..", firefoxDir, "lib", "main.js"), 'w') as content_file:
		content_file.write(main)

	print("Firefox SDK")
	os.chdir(os.path.join("..", sdkDir))
	sdkRoot = os.getcwd()
	sys.path.append(os.path.join(sdkRoot, "python-lib"))
	os.environ["CUDDLEFISH_ROOT"] = sdkRoot

	from jetpack_sdk_env import welcome
	welcome()

	print("Export .xpi")
	import cuddlefish

	os.chdir(os.path.join("..", firefoxDir))
	#try :
	#	cuddlefish.run(["--force-mobile", "--update-url", "https://thetabx.net/addon/tga/checkfirefoxupdate/%APP_OS%/%CURRENT_APP_VERSION%/%ITEM_VERSION%/", "xpi"])
	#except SystemExit:
	#	print("Copy mobile .xpi")
	#	distutils.file_util.copy_file("gksimproved.xpi", os.path.join("..", chromeDir, "build", "gksimproved.mobile.xpi"))

	try :
		cuddlefish.run(["--update-url", "https://thetabx.net/addon/tga/firefox/%APP_OS%/%CURRENT_APP_VERSION%/%ITEM_VERSION%/", "xpi"])
	except SystemExit:
		print("Copy .xpi")
		distutils.file_util.copy_file("tgarmory.xpi", os.path.join("..", chromeDir, "build"))

	os.chdir(os.path.join("..", chromeDir))
	print("Done")

File: tools/buildFirefox.js
var fs = require('fs-extra');
var path = require('path');
var exec = require('child_process').exec;

function cleanUp() {
  console.log('Firefox : Cleanups');
  fs.removeSync(path.join('..', 'firefox'));
  fs.mkdirsSync(path.join('..', 'firefox'));
}

function copyFiles() {
  console.log('Firefox : Copy files');
  var copyDirs = ['images', 'lib', 'css'];
  copyDirs.forEach(function(d) {
    fs.copySync(path.join('..', d), path.join('..', 'firefox', d));
  });
  var filesReg = new RegExp(/\.(js|json)?$/);
  var copyFiles = fs.readdirSync(path.join('..')).filter(function(f) {
    return filesReg.test(f);
  });
  copyFiles.forEach(function(f) {
    fs.copySync(path.join('..', f), path.join('..', 'firefox', f));
  });
}

function updateManifest() {
  console.log('Firefox : Update manifest');
  var manifest = fs.readJsonSync(path.join('..', 'firefox', 'manifest.json'));
  manifest.applications = {
    gecko: {
      id: 'tgarmory@thetabx.net',
      strict_min_version: '48.0',
      update_url: 'https://thetabx.net/addon/tgarmory/firefox/%APP_OS%/%CURRENT_APP_VERSION%/%ITEM_VERSION%/',
    }
  };
  fs.writeJsonSync(path.join('..', 'firefox', 'manifest.json'), manifest);
}

function doBuild(cb) {
  console.log('Firefox : Build');
  process.chdir(path.join('..', 'firefox'));
  var firefoxCreds = fs.readJsonSync(path.join('..', 'tools', 'firefox.creds'));
  var cmd = ['web-ext', 'sign', '--api-key', firefoxCreds.user, '--api-secret',
    firefoxCreds.secret
  ];
  exec(cmd.join(' '), function(error, stdout, stderr) {
    if (stdout) {
      console.log(stdout);
    }
    if (error) {
      console.error(error);
      return;
    }
    if (stderr) {
      console.error(stderr);
      return;
    }
    console.log('Success');
    if (cb) {
      cb();
    }
  });
}

var xpiReg = new RegExp(/\.xpi?$/);
function moveXpi(cb) {
  console.log('Firefox : Move');
  var artifactDir = 'web-ext-artifacts';
  var xpis = fs.readdirSync(artifactDir).filter(function(f) {
    return xpiReg.test(f);
  });
  xpis.forEach(function(xpi) {
    fs.move(path.join(artifactDir, xpi), path.join('..', 'build', 'tgarmory.xpi'), {
      clobber: true
    }, function(e) {
      if (e) {
        console.error(e);
      } else {
        cb();
      }
    });
  });
}

function build(cb) {
  console.log('Build firefox');
  cleanUp();
  copyFiles();
  updateManifest();
  doBuild(function() {
    moveXpi(cb);
  });
}

exports.build = build;
